### ISSUE TITLE
[question] deterministic build and source dir 

### ISSUE THREAD

**Issue by boussaffawalid** [conan-io/conan#11245]
I finally manage to achieve 100% deterministic builds. One of the issue was randomness in conan build and source directories when creating conan packages. 

The solution I used with the [old cmake build helper  ](https://docs.conan.io/en/1.43/reference/build_helpers/cmake.html) was to create directory junction and pass them to cmake instead of the conan source and build folders: 

```
mklink /J x:\ci\source\ self.source_folder
mklink /J x:\ci\build\ self.build_folder
```

```
cmake.configure(source_dir="x:\ci\source\", build_dir="x:\ci\build\")
```

Now I switched to the new conan helpers https://docs.conan.io/en/latest/reference/conanfile/tools/cmake.html and it is not possible anymore to override source_dir and build_dir. 

I tried using the layout method

```
def layout(self):
    self.folders.source = "x:\ci\source\"
    self.folders.build = "x:\ci\build\"
```
but I got blocked by source_dir not available in the layout method.

Any advices on how to achieve deterministic build and source dirs in conan.

**Comment by memsharded** [comment]
Using an absolute path for all packages seems a bit fragile, IMHO.

Then I'd like to understand that "randomness", because Conan cache folders are deterministic in 1.X (for the same package, they build in the same folder always). What am I missing?

**Comment by boussaffawalid** [comment]
Build and source folder information are propagated to binaries because of many reasons (usage of `__FILE__` macro for example).  
When rebuilding the same source code but in different location, you get different binaries. And because conan use different paths for different versions, final binaries are different even if source code did not change.

**Comment by memsharded** [comment]
> Build and source folder information are propagated to binaries because of many reasons (usage of __FILE__ macro for example).
When rebuilding the same source code but in different location, you get different binaries. 

Sure, we know about this :) https://blog.conan.io/2019/09/02/Deterministic-builds-with-C-C++.html

The question is why do you need to do that. If you are using Conan 1.X, as long as you put the Conan cache in the same place (lets say, ``C:/myconan``, you will also get exactly the same path, like ``.../data/pkg/version/user/channel/package_id/build``, this path is fully deterministic in Conan 1.X.


The second question is regarding: 

> but I got blocked by source_dir not available in the layout method.

What do you mean? No idea what ``source_dir`` in the ``layout()`` method means, yes, in that method you get ``self.folders.xxx`` definitions, but you can define that there.

**Comment by boussaffawalid** [comment]
In the path  `.../data/pkg/version/user/channel/package_id/build`, `version` and `package_id` could change without changes in the c++ source code.
For example, in our case, we generate a different a different package version for every commit. so the path `.../data/pkg/version/user/channel/package_id/build` is never the same.

**Comment by memsharded** [comment]
> For example, in our case, we generate a different a different package version for every commit. 

Why would you bump the version if you are not doing changes at all to the source code? This is a bit surprising to me, just trying to understand your flow and constraints.

**Comment by boussaffawalid** [comment]
no changes in c++ code I mean.

Let's say somebody update a readme file, commit and push changes, in our setup, we create a new conan package with a new conan version, that means conan will now build in a different directory. 

Now, the hashes of all c++ objects (lib, dll, exe) are different because the build and source directory are different.  (because some c++ file uses __FILE__ macro for example).

**Comment by memsharded** [comment]
> Let's say somebody update a readme file, commit and push changes, in our setup, we create a new conan package with a new conan version, that means conan will now build in a different directory.

I see. This is one of the things that ``revisions`` aim to reduce. For example, if using ``revisions``, and ``exports_sources``, then the recipe revision will not be modified, and it will even know that it doesn't need to rebuild anything, no wasting an extra version and extra build time. Revisions were basically born as a way to track changes to things different to the C++ source code, that would require a version bump, like when tracking versions of open source libraries, but still doing some changes to the recipe. It is true that if using other methods as ``scm`` based, you get a new revision. In Conan 1.X, new revisions still get the same build directory (but that might change in Conan 2.0 with the multi-revision enabled cache).

~So now, I would like to know more details what is failing for you in the ``layout()`` approach. If you can provide more details about what you tried, and what issues are you seeing, that would help.~

Update: Now I think I see the issue. The ``layout()`` is intended to describe the relative location of things, but it is not a "use this location instead" mechanism.

**Comment by boussaffawalid** [comment]
we use the `scm` method as we keep the recipe in our source control for internal products. btw, We need deterministic builds for internal projects only as for we rarely rebuild third party dependency.

in theory I want to achieve something like this.
```
def layout(self):
    self.run("mklink /J x:\ci\source\ self.source_folder")
    self.run("mklink /J x:\ci\build\ self.build_folder")

    self.folders.source = "x:\ci\source\"
    self.folders.build = "x:\ci\build\"
```

Basically fooling cmake to always use `x:\ci\source\` as source directory and `x:\ci\build\` as build directory.

the issue now is that I cannot run the mklink command in the layout method, as it does not make sense, since `self.source_folder` and `self.build_folder` are not defined yet. 

I think I found a solution for build_folder: adding a pre_build hook that:
* `mklink /J x:\ci\build\ self.build_folder`
* update `binaryDir` in cmake `CMakePresets.json` and set it to `x:\ci\build\`

Now, still looking for a solution for source dir

**Comment by memsharded** [comment]
Quick question, why don't you move the ``mklink`` to the start of the ``build()`` method directly?

**Comment by boussaffawalid** [comment]
well, I want to use a pre_build hook just to keep the recipe clean, but `mklink` can be moved to the `build` method as well.
The question now, how to fool [cmake helper](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmake.html) to use `x:\ci\source\` instead of `self.source_folder`

from conan docs,  [cmake constructor](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmake.html#constructor), 

> build_script_folder: Relative path to the folder containing the root CMakeLists.txt

Not sure if this supports absolute path but I will give it a try and set it to `x:\ci\source\` and see how it goes

**Comment by jcar87** [comment]
Hi @boussaffawalid, thanks for asking this question - concerns about reproducibility are raised by the community from time to time.

I'm assuming that you're on Windows using the `msvc` compiler - In this case, I believe the underlying issue is that this compiler has some sources of non-determinism, and it does not provide flags to override or replace the paths, thus leaving you with slightly different generated binaries. 

In particular, msvc is likely to embed the full path to the compiled sources wherever the `__FILE__` macro is used. Recent versions of `gcc` and `clang` have a handy flag to "remap" the path to something else - which would allow us to achieve repeatable builds (at least with regards to this macro). 
You can see an example comparing the generated assembly between `msvc` and `gcc` (with the `-fmacro-prefix-map` flag) [here](https://godbolt.org/z/81ebdYvba).

Unfortunately, unless VC++ provides a similar flag (to `macro-prefix-map`), there isn't much that can be done, other than to patch the binaries after the fact. There are other macros defined by the compiler that might affect reproducibility if they are used, for example `__DATE__` and `__TIME__` (see [here](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170)).

In your case it might be possible to patch the files after the fact, in particular:

> For example, in our case, we generate a different a different package version for every commit. so the path .../data/pkg/version/user/channel/package_id/build is never the same.


If for every commit the package_id is changing - the length of the package ID in conan will remain the sabe, so it is possible to patch this afterwards - there are some pointers near the end of[ _this blog post_ ](https://blog.conan.io/2019/09/02/Deterministic-builds-with-C-C++.html)on how to use hooks to achieve this.

> We need deterministic builds for internal projects only as for we rarely rebuild third party dependency.

If you are only using this for internal projects, and this is also a CMake project - you may wish to try alternatives such as using cmake to pass a compiler define specific to each source file: https://cmake.org/pipermail/cmake/2011-December/048281.html, and avoid using `__FILE__` altogether.

At the end of the day, we're dealing with a macro for which compilers follow their own convention, and a compiler that doesn't provide any mechanism to override this. I would also recommended voting for the[ outstanding request] that would address this issue (https://developercommunity.visualstudio.com/t/map-file-to-a-relative-path/1393881) at VS Developer Community.

**Comment by memsharded** [comment]
To complete with further suggestions, specially moving forward to Conan 2.0:

- We have just merged in Conan 2.0 #11447, to guarantee that the build folder for a given recipe, version, recipe revision and package_id will always be the same. Together with setting a common CONAN_HOME environment variable, can guarantee the exact same folder always, for a given recipe, version, revision and package_id
- If even 2 different versions must guarantee the exact same binary byte to byte, that is outside Conan scope. There are a few different approaches that could be followed:
  - Fixing things at the build level, as commented above. Request MSVC to provide equivalent functionality to other compilers to make builds location independent
  - Do not build in the Conan cache. Use local commands, ``conan build`` + ``conan export-pkg`` in the folder of your convenience to create packages. If you want to further automate this in a single command, Conan custom commands new 2.0 feature will allow creating and sharing your own commands easily.

**Comment by AbrilRBS** [comment]
Closing as answered — the practical need here is already covered: Conan 2.0 guarantees deterministic cache folders per recipe/version/revision/package_id (#11447) when combined with a fixed `CONAN_HOME`. The remaining ask — byte-identical binaries across genuinely different versions — was explicitly scoped as outside Conan's job, with two concrete alternatives (fix at the compiler/build level, or build outside the cache with `conan build` + `export-pkg`) offered instead.

**Comment by memsharded** [comment]
> Closing as answered — the practical need here is already covered: Conan 2.0 guarantees deterministic cache folders per recipe/version/revision/package_id (https://github.com/conan-io/conan/pull/11447) when combined with a fixed CONAN_HOME. The remaining ask — byte-identical binaries across genuinely different versions — was explicitly scoped as outside Conan's job, with two concrete alternatives (fix at the compiler/build level, or build outside the cache with conan build + export-pkg) offered instead.

I am afraid this is still not possible. It was necessary to add randomness to the build folders, because using the same folder was not possible, because the built binaries cannot be moved to a permanent folder due to the issues of copying those binaries around (like Windows Antivirus blocking the copy and failing, etc)

Let's re-open, I think the deterministic folders is still something that is not possible in the Conan cache.
Users can still use the local flow, ``conan source + conan build + conan export-pkg`` and sure they can automate things with a custom command too. What about building something from Artifactory like a dependency? With ``workspace open`` it might be possible to do that.

So lets at least document the recommendations for people needing to do this, it might be a bit niche requirement, but it might be worth properly documenting some recommendations.
